pkg/modules/eventprocessing.go

```diff
@@ -0,0 +1,122 @@
+package modules
+
+import (
+	"fmt"
+	"github.com/filecoin-project/mir/pkg/events"
+	"github.com/filecoin-project/mir/pkg/pb/eventpb"
+	"runtime/debug"
+)
+
+// ApplyEventsSequentially takes a list of events and applies the given applyEvent function to each event in the list.
+// Processing is performed sequentially, one event at a time, in the order of the input list.
+// The EventLists returned by applyEvent are aggregated in a single EventList (in order of creation)
+// and returned by ApplyEventsSequentially.
+func ApplyEventsSequentially(
+	eventsIn *events.EventList,
+	applyEvent func(*eventpb.Event) (*events.EventList, error),
+) (*events.EventList, error) {
+
+	eventsOut := &events.EventList{}
+
+	iter := eventsIn.Iterator()
+	for event := iter.Next(); event != nil; event = iter.Next() {
+		evts, err := applyEvent(event)
+		if err != nil {
+			return nil, err
+		}
+		eventsOut.PushBackList(evts)
+	}
+
+	return eventsOut, nil
+}
+
+// ApplyEventsConcurrently takes a list of events and applies the given applyEvent function to each event in the list.
+// Processing is performed concurrently on all events in the input list.
+// Thus, the provided applyEvent function must be thread-safe.
+// The EventLists returned by applyEvent are aggregated in a single EventList and returned by ApplyEventsSequentially.
+// Despite being executed concurrently,
+// the order of the returned results preserves the order of the corresponding input events.
+// Thus, if applyEvent is deterministic, the output of ApplyEventsConcurrently is also deterministic.
+// If one or more errors occur during processing, ApplyEventsConcurrently returns the first of them,
+// along with an empty EventList.
+func ApplyEventsConcurrently(
+	eventsIn *events.EventList,
+	applyEvent func(*eventpb.Event) (*events.EventList, error),
+) (*events.EventList, error) {
+
+	// Initialize channels into which the results of each invocation of applyEvent will be written.
+	results := make([]chan *events.EventList, eventsIn.Len())
+	errors := make([]chan error, eventsIn.Len())
+	for i := 0; i < eventsIn.Len(); i++ {
+		results[i] = make(chan *events.EventList)
+		errors[i] = make(chan error)
+	}
+
+	// Start one concurrent worker for each event in the input list.
+	// Note that the processing starts concurrently for all events, and only the writing of the results is synchronized.
+	iter := eventsIn.Iterator()
+	i := 0
```

Comment:
I think this variable is no longer used.

Reply:
The `i` variable? It is, as each goroutine needs to know the index of its output channels and the `Iterator` implementation does not provide that information.